src/MKRWAN.h

```diff
@@ -1047,79 +986,135 @@ class LoRaModem : public Stream
                        ConstStr r1=GFP(LORA_OK), ConstStr r2=GFP(LORA_ERROR),
-                       ConstStr r3=NULL, ConstStr r4=NULL, ConstStr r5=NULL,
-                       ConstStr r6=NULL, ConstStr r7=NULL, ConstStr r8=NULL)
+                       ConstStr r3=GFP(LORA_ERROR_PARAM), ConstStr r4=GFP(LORA_ERROR_BUSY), ConstStr r5=GFP(LORA_ERROR_OVERFLOW),
+                       ConstStr r6=GFP(LORA_ERROR_NO_NETWORK), ConstStr r7=GFP(LORA_ERROR_RX), ConstStr r8=GFP(LORA_ERROR_UNKNOWN))
   {
-    data.reserve(64);
+    data.reserve(msize);
     int8_t index = -1;
     int length = 0;
+    int a = 0;
     unsigned long startMillis = millis();
     do {
       YIELD();
       while (stream.available() > 0) {
-        int a = streamRead();
+        a = stream.peek();
         if (a < 0) continue;
-        data += (char)a;
-        if (r1 && data.endsWith(r1)) {
-          index = 1;
-          goto finish;
-        } else if (r2 && data.endsWith(r2)) {
-          index = 2;
-          goto finish;
-        } else if (r3 && data.endsWith(r3)) {
-          index = 3;
-          goto finish;
-        } else if (r4 && data.endsWith(r4)) {
-          index = 4;
-          goto finish;
-        } else if (r5 && data.endsWith(r5)) {
-          index = 5;
-          goto finish;
-        } else if (r6 && data.endsWith(r6)) {
-          index = 6;
-          goto finish;
-        } else if (r7 && data.endsWith(r7)) {
-          index = 7;
-          goto finish;
-        } else if (r8 && data.endsWith(r8)) {
-          index = 8;
-          goto finish;
-        } else if (data.endsWith("+RECV=")) {
-          data = "";
-          stream.readStringUntil(',').toInt();
-          length = stream.readStringUntil('\r').toInt();
-          streamSkipUntil('\n');
-          streamSkipUntil('\n');
-          for (int i = 0; i < length;) {
-            if (stream.available()) {
-                rx.put(stream.read());
-                i++;
-            }
-          }
+        if (a == '=' || a == '\r'
+        		|| (a == '+' && data.length() > 0)) {
+			DBG("### Data string:", data);
+			if (r1 && data.endsWith(r1)) {
+			  index = 1;
+			  goto finish;
+			} else if (r2 && data.endsWith(r2)) {
+			  index = 2;
+			  goto finish;
+			} else if (r3 && data.endsWith(r3)) {
+			  index = 3;
+			  goto finish;
+			} else if (r4 && data.endsWith(r4)) {
+			  index = 4;
+			  goto finish;
+			} else if (r5 && data.endsWith(r5)) {
+			  index = 5;
+			  goto finish;
+			} else if (r6 && data.endsWith(r6)) {
+			  index = 6;
+			  goto finish;
+			} else if (r7 && data.endsWith(r7)) {
+			  index = 7;
+			  goto finish;
+			} else if (r8 && data.endsWith(r8)) {
+			  index = 8;
+			  goto finish;
+			} else if ((data.endsWith(AT_RECV)
+					|| data.endsWith(AT_RECVB)) && a == '=') {
+			  (void)stream.readStringUntil(',').toInt();
+			  length = stream.readStringUntil('\r').toInt();
+			  (void)streamSkipUntil('\n');
+			  (void)streamSkipUntil('\n');
```

Comment:
index should be 99 here?

Reply:
The comment says so, but it seems the return code 99 has been removed in favor of buffer parsing + return state evaluation. For example, you have the modem in Class C. If you set a value and read from the modem, it could respond with `+RECV=<somemessage>+OK`. In this case, it pastes data to the buffer but still returns ok to the parameter set. You can poll if some new data arrived. If we would return 99, we get an error for the set as we didn't read the expected `+OK` yet, or we prefer the _receive_ to the ok.